elastictransport/instrumentation.go

```diff
@@ -0,0 +1,215 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package elastictransport
+
+import (
+	"bytes"
+	"context"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+	"io"
+	"net/http"
+	"strconv"
+)
+
+const schemaUrl = "https://opentelemetry.io/schemas/1.21.0"
+const tracerName = "elasticsearch-api"
+
+// Constants for Semantic Convention
+// see https://opentelemetry.io/docs/specs/semconv/database/elasticsearch/ for details.
+const attrDbSystem = "db.system"
+const attrDbStatement = "db.statement"
+const attrDbOperation = "db.operation"
+const attrDbElasticsearchClusterName = "db.elasticsearch.cluster.name"
+const attrDbElasticsearchNodeName = "db.elasticsearch.node.name"
+const attrHttpRequestMethod = "http.request.method"
+const attrUrlFull = "url.full"
+const attrServerAddress = "server.address"
+const attrServerPort = "server.port"
+const attrPathParts = "db.elasticsearch.path_parts."
+
+// Instrumentation defines the interface the client uses to propagate information about the requests.
+// Each method is called with the current context or request for propagation.
+type Instrumentation interface {
+	// Start creates the span before building the request, returned context will be propagated to the request by the client.
+	Start(ctx context.Context, name string) context.Context
+
+	// Close will be called once the client has returned.
+	Close(ctx context.Context)
+
+	// RecordError propagates an error.
+	RecordError(ctx context.Context, err error)
+
+	// RecordPathPart provides the path variables, called once per variable in the url.
+	RecordPathPart(ctx context.Context, pathPart, value string)
+
+	// RecordQuery provides the endpoint name as well as the current request payload.
+	RecordQuery(ctx context.Context, endpoint string, query io.Reader) io.ReadCloser
```

Comment:
Should that be `RecordRequestBody`? The term `Query` is generally used for the URL query string.

Reply:
An excellent suggestion, done.